libs/parser/parser.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * Phoenix-Pilot
+ *
+ * Parser for config files
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Piotr Nieciecki
+ *
+ * This file is part of Phoenix-Pilot software
+ *
+ * %LICENSE%
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+
+#include "parser.h"
+
+#define MAX_HEADER_LEN 16
+#define MAX_FIELD_LEN  16
+#define MAX_VALUE_LEN  64
+
+#define WORD "[A-Za-z0-9_.,-]"
+
+
+/*
+ * Patterns for regex expressions
+ *
+ * If you have changed these patterns check if these function work properly:
+ * - parser_nextHeaderGet
+ * - parser_nextFieldNameGet
+ * - parser_nextFieldGet
+ *
+ * Especially be aware of `regmatch_t` structures.
+ */
+#define HEADER_PATTERN        "^[[:space:]]*@(" WORD "+)[[:space:]]*(#.*)?[[:space:]]*$"
+#define HEADER_SUBEXPRESSIONS 2
+
+#define FIELD_PATTERN        "^[[:space:]]*(" WORD "+)[ =]+(" WORD "+)[[:space:]]+(#.*)?[[:space:]]*$"
+#define FIELD_SUBEXPRESSIONS 3
+
+#define COMMENT_PATTERN "^[[:space:]]*#.*$|^[[:space:]]*$"
+
+
+/* Static variables */
+struct {
+	int refs;
+
+	regex_t headerRegex;
+	regex_t fieldRegex;
+	regex_t commentRegex;
+} parser_common;
+
+
+/* Structure contains information about defined header */
+typedef struct {
+	char headerName[MAX_HEADER_LEN + 1];
+	int (*converter)(hmap_t *);
+} parser_headerInfo_t;
+
+
+/* Contains information about particular field from file */
+typedef struct {
+	char field[MAX_FIELD_LEN + 1];
+	char value[MAX_VALUE_LEN + 1];
+} parser_field_t;
+
+
+/* Structure of the main parser object */
+struct parser_t {
+	parser_headerInfo_t *headerInfos;
+	hmap_t *headersMap;
+
+	parser_field_t *fields;
+	hmap_t *fieldsMap;
+};
+
+
+enum parser_lineType { lineType_Header, lineType_Field, lineType_Comment, lineType_InvalidLine };
```

Comment:
Meeting enum naming convention (+shorter):
```suggestion
enum parser_lineType { type_header, type_field, type_comment, type_invalid };
```